### PR TITLE
store: thread safe for Register kv.Driver

### LINKED ISSUE #28318: Unstable test store_test.go fatal error: concurrent map writes
```
[2021-09-24T16:52:41.182Z] fatal error: concurrent map writes
[2021-09-24T16:52:41.182Z] 
[2021-09-24T16:52:41.182Z] goroutine 1416 [running]:
[2021-09-24T16:52:41.182Z] runtime.throw(0x3bbc016, 0x15)
[2021-09-24T16:52:41.182Z] 	/usr/local/go/src/runtime/panic.go:1117 +0x72 fp=0xc0067e2678 sp=0xc0067e2648 pc=0x1232e72
[2021-09-24T16:52:41.182Z] runtime.mapassign_faststr(0x3741280, 0xc0003dcd80, 0x3b8de6f, 0x5, 0x5e668e0)
[2021-09-24T16:52:41.182Z] 	/usr/local/go/src/runtime/map_faststr.go:211 +0x3f1 fp=0xc0067e26e0 sp=0xc0067e2678 pc=0x120f131
[2021-09-24T16:52:41.182Z] github.com/pingcap/tidb/store.Register(0x3b8de6f, 0x5, 0x40568e0, 0x5e64400, 0x614e01bd, 0xb5a755a)
[2021-09-24T16:52:41.182Z] 	/home/jenkins/agent/workspace/tidb_ghpr_check_2/go/src/github.com/pingcap/tidb/store/store.go:38 +0xb6 fp=0xc0067e2738 sp=0xc0067e26e0 pc=0x34d1dd6
[2021-09-24T16:52:41.182Z] github.com/pingcap/tidb/store.TestRegister(0xc005de7980)
[2021-09-24T16:52:41.182Z] 	/home/jenkins/agent/workspace/tidb_ghpr_check_2/go/src/github.com/pingcap/tidb/store/store_test.go:794 +0x65 fp=0xc0067e2780 sp=0xc0067e2738 pc=0x34dbd05
[2021-09-24T16:52:41.182Z] testing.tRunner(0xc005de7980, 0x3c8f528)
[2021-09-24T16:52:41.182Z] 	/usr/local/go/src/testing/testing.go:1193 +0xef fp=0xc0067e27d0 sp=0xc0067e2780 pc=0x1327e0f
[2021-09-24T16:52:41.182Z] runtime.goexit()
[2021-09-24T16:52:41.182Z] 	/usr/local/go/src/runtime/asm_amd64.s:1371 +0x1 fp=0xc0067e27d8 sp=0xc0067e27d0 pc=0x126da61
[2021-09-24T16:52:41.182Z] created by testing.(*T).Run
[2021-09-24T16:52:41.182Z] 	/usr/local/go/src/testing/testing.go:1238 +0x2b3
```

### PR DESCRIPTION
**Pull request by w169q169**
### What problem does this PR solve?

Issue Number: close #28318

introduced in #27993

Problem Summary:

### What is changed and how it works?

What's Changed:
`Register` function is not thread safe.
https://github.com/pingcap/tidb/blob/e60f20b9f4ee502bec4cebd84de18a25fefeebb4/store/store.go#L31-L40

Run those  parallel test functions will suffer fatal error for concurrent map writes on occasionally.
https://github.com/pingcap/tidb/blob/e60f20b9f4ee502bec4cebd84de18a25fefeebb4/store/store_test.go#L765-L770
https://github.com/pingcap/tidb/blob/e60f20b9f4ee502bec4cebd84de18a25fefeebb4/store/store_test.go#L780-L785
How it Works:
Add mutex Lock for map variable in package write and read.

### Release note 

```release-note
None
```

### CODE REVIEW

**Comment by w169q169** [comment]
/cc @tisonkun @morgo 
PTAL

**Review by tisonkun — COMMENTED** [review]
I'd prefer move the two tests touching "Register" in serial tests.

I'm unsure whether or not the function should be design as thread safe or with assumption only call in serial. But if you can elaborate the reason, I can agree with this change.

**Comment by tisonkun** [comment]
Also you can report an unstable test in #25899 (if you don't have write permission, create an unstable test issue and notify me) and let this PR fixes it.

**Comment by w169q169** [comment]
> I'd prefer move the two tests touching "Register" in serial tests.
> 
> I'm unsure whether or not the function should be design as thread safe or with assumption only call in serial. But if you can elaborate the reason, I can agree with this change.

moving the tests using `Register` in serial tests is quit a litter easy to fix the problem.
`Register` only run parallel in test scenes, howerver if people use the function not too carefully  it will cause a thread error that is unstable and unnotice.So we can use the defensesive code  (suppose that people always use wrong way ) to avoid the error will happen in the future ,  even on current the function is used right in most scenes.
if you have diffent idea, please let me know and I will fix the test according to your opinion.

**Review by tisonkun — APPROVED** [review]
You convince me that this function is not in a significant path and we can use defensive code trade off a bit performance.

@morgo please give a final check.

... also, I'm curious that Golang doesn't have a concurrent map out-of-the-box?

File: store/store.go [hunk 1]
```diff
@@ -59,11 +64,17 @@ func newStoreWithRetry(path string, maxRetries int) (kv.Storage, error) {
 	}
 
 	name := strings.ToLower(storeURL.Scheme)
+
+	storesLock.RLock()
+
 	d, ok := stores[name]
```
Comment (tisonkun) [inline store/store.go:70]: Move two `storesLock.RUnlock()` right after this line?

**Review by tisonkun — CHANGES_REQUESTED** [review]
Can we instead use an existing concurrent map struct or lib? The diff itself is a bit handy and brittle.

Both two changes are not as narrow or clean, or panic safe. We should have written thread-safe read/write or used an existing one.

File: store/store.go (hunk 1 above)
Comment (tisonkun) [reply store/store.go:70]: ... or write a function to retrieve the driver, like:

```go
func loadDriver(name string) (kv.Driver, bool) {
	storesLock.RUnlock()
	defer storesLock.Unlock()
	d, ok := stores[name]
	return d, ok
}
```

Comment (w169q169) [reply store/store.go]: I'd prefer write a function to get the driver , and scope for lock is more less in this way

**Review by tisonkun — APPROVED** [review]
LGTM.

**Comment by tisonkun** [comment]
/cc @morgo @disksing

PTAL.

**Comment by disksing** [comment]
> /cc @morgo @disksing
> 
> PTAL.

LGTM. Defer to @morgo for another check.

**Comment by tisonkun** [comment]
@morgo this unstable test occurs several times in these days. Could you please do a final check on this fix?

**Review by morgo — APPROVED** [review]

**Comment by morgo** [comment]
Thank you for the contribution!

**Pull request merged**